File: go/debug.go
package main

import (
	"fmt"
	"os"
	"path/filepath"
	"sync"
	"time"
)

type DebugLogger struct {
	mu   sync.Mutex
	file *os.File
}

func NewDebugLogger(path string) *DebugLogger {
	if path == "" {
		path = filepath.Join(os.TempDir(), fmt.Sprintf("slit-%d.log", os.Getpid()))
	}

	f, err := os.OpenFile(path, os.O_CREATE|os.O_WRONLY|os.O_APPEND, 0644)
	if err != nil {
		return nil
	}

	return &DebugLogger{file: f}
}

func (dl *DebugLogger) Log(format string, args ...interface{}) {
	if dl == nil || dl.file == nil {
		return
	}
	dl.mu.Lock()
	defer dl.mu.Unlock()
	ts := time.Now().Format("15:04:05.000")
	msg := fmt.Sprintf(format, args...)
	fmt.Fprintf(dl.file, "%s %s\n", ts, msg)
}

func (dl *DebugLogger) Close() {
	if dl == nil || dl.file == nil {
		return
	}
	dl.file.Close()
	dl.file = nil
}

File: rust/src/debug.rs
use std::fs::File;
use std::io::Write;
use std::sync::Mutex;

pub struct DebugLogger {
    file: Mutex<Option<File>>,
}

impl DebugLogger {
    pub fn new(path: Option<&std::path::Path>) -> Option<Self> {
        let resolved = path.map(|p| p.to_path_buf()).unwrap_or_else(|| {
            std::env::temp_dir().join(format!("slit-{}.log", std::process::id()))
        });
        match File::options().create(true).append(true).open(&resolved) {
            Ok(f) => Some(Self {
                file: Mutex::new(Some(f)),
            }),
            Err(_) => None,
        }
    }

    pub fn log(&self, msg: &str) {
        if let Ok(mut guard) = self.file.lock() {
            if let Some(ref mut f) = *guard {
                let now = chrono::Local::now();
                let _ = writeln!(f, "{} {}", now.format("%H:%M:%S%.3f"), msg);
            }
        }
    }

    pub fn close(&self) {
        if let Ok(mut guard) = self.file.lock() {
            *guard = None;
        }
    }
}
